### ISSUE TITLE
ConnLimit load-balancer visualisation

### ISSUE THREAD

**Issue by mihaitodor** [traefik/traefik#1650]
### Do you want to request a *feature* or report a *bug*?

feature

### Description

Currently, the Traefik REST API and UI do not expose the number of active connections and the total number of connections when the ConnLimit load-balancer is enabled for (some) backends.

In this sense, I wrote [some code](https://github.com/mihaitodor/traefik/commits/backend-conn-stats) which exposes these numbers through the REST API for each backend that has ConnLimit enabled and I added some UI graphs which help us visualise the Historical Connection Counts through a line chart and the Current Connection Counts (versus the max counts) through a discrete bar chart. Here is a preview of the UI additions:

![](https://media.giphy.com/media/xUPGcITQd0LHupfPYk/giphy.gif)

In order to make this work, I needed access to some [private data](https://github.com/containous/oxy/blob/f88530866c561d24a6b5aac49f76d6351b788b9f/connlimit/connlimit.go#L19) from your fork of [vulcand/oxy](https://github.com/containous/oxy), which I extracted via a reflection hack [here](https://github.com/mihaitodor/traefik/commit/ff1302dc436ddb9467b74de8d0ffa29589b222c1#diff-a3520fb65206c9e232f10baaff310c7aR394). If you are interested in maintaining extra changes to this oxy fork, I'm more than happy to push a PR to it, in order to not need reflection for extracting the connection counts. This shouldn't have any impact on the performance, so I don't mind keeping the reflection code.

Regarding the REST API, I created 2 new routes: [`api/conn_stats`](https://github.com/mihaitodor/traefik/commit/ff1302dc436ddb9467b74de8d0ffa29589b222c1#diff-a3520fb65206c9e232f10baaff310c7aR128) and [`api/providers/{provider}/backends/{backend}/conn_stats`](https://github.com/mihaitodor/traefik/commit/ff1302dc436ddb9467b74de8d0ffa29589b222c1#diff-a3520fb65206c9e232f10baaff310c7aR129). The UI additions [require only the first route](https://github.com/mihaitodor/traefik/commit/77b1e06a0683f5e411efe22db1097333c5176ab5#diff-74137107f6bef14865fd3a0d9740d950R13), but I think it might be handy to have the second one as well, in case somebody else wants to implement some sort of drill-down functionality in the UI (or some 3rd party plugin).

Please let me know what you think and I will rebase my changes and create a PR if you find this feature useful.

**Comment by timoreimann** [comment]
@mihaitodor this looks quite useful on first sight.

One question: Is it possible to implement this feature without relying on the connection limiter being enabled? That would open up the metrics to users that don't need the limiter (and possibly other parts of Traefik -- off the top of my head, I can think of metrics).

**Comment by mihaitodor** [comment]
> That would open up the metrics to users that don't need the limiter 

Yes, I agree, and I have considered that, but that would mean that everyone needs to pay the price of [acquiring a mutex](https://github.com/containous/oxy/blob/f88530866c561d24a6b5aac49f76d6351b788b9f/connlimit/connlimit.go#L75) for each new connection, which, depending on the use case, might not be acceptable. I assumed this is the reason why ConnLimits is an optional middleware that is not enabled by default.

PS: I'm open to more ideas about adding more charts related to connections. I saw that the Prometheus middleware exposes some sort of [latency histogram](https://github.com/containous/traefik/blob/12a0026e211eaca9d0244922421bde64acae5f32/middlewares/prometheus.go#L21) that I haven't studied in detail yet.

**Comment by timoreimann** [comment]
@mihaitodor in general, Go mutexes are implemented quite efficiently. But you're right that it's still a performance question, something that should likely be measured accurately. Personally, I'm okay with starting with the safe route of coupling to ConnLimits. 👍 

@emilevauge should have a look on the feature request too.

**Comment by mihaitodor** [comment]
@emilevauge Please let me know if you have any feedback regarding this feature.

**Comment by emilevauge** [comment]
Hey @mihaitodor, not sure to understand the link between this feature and the need of using a connection limiter (and also mutex). Is it possible to not use this middleware at all?

**Comment by mihaitodor** [comment]
@emilevauge OK, so @timoreimann expressed the same concern above. The issue is that, by default, Traefik doesn't seem to keep track of the connection count per backend (please correct me if I'm wrong). The connection limiter middleware has an implementation for this ([via a mutex](https://github.com/containous/oxy/blob/f88530866c561d24a6b5aac49f76d6351b788b9f/connlimit/connlimit.go#L75)) and I decided to leverage it for this feature because at Nitro we always want to have the connection limiter middleware enabled.

I could implement the required backend functionality in some other place (I'm open to suggestions here) and have that enabled by default, but then we might have to consider the cost of counting connections (via a mutex or some atomic increment / decrement - no clue which is more efficient). Also, it could annoy people who enable the connection limiter middleware, because they'd have to pay the price of a mutex twice. Not sure how much overhead this adds, though, but I can imagine some people seeing a performance degradation.

**Comment by emilevauge** [comment]
@mihaitodor OK got it. Thanks for the explanation.
I would prefer to implement a specific middleware to keep track of connections count by backend. But I also think it should be possible to not use any mutex to achieve this (for example, using a infinite channel).
@timoreimann any thoughts on this ?

**Comment by timoreimann** [comment]
I noticed we have this [StatsRecorder](https://github.com/containous/traefik/blob/885b9f371c08e60c055f236a0664f0279d9d2786/middlewares/stats.go) middleware. I'm not really familiar with it, but it already uses a mutex -- could we collect the metrics we need within this context possibly?

If not or when in doubt about the performance aspect, I'd follow Knuth's advice and not worry too much about the performance aspect for now. My suggestion would be to work towards having the functionality live nicely in a dedicated middleware, measure performance, and optimize if (and only if) we need to. My gut feeling tells me that another mutex wouldn't pose a major concern; and even if it had an effect, users would still make a deliberate choice to enable the feature at a known cost.

Re: go-kit/Prometheus: I'm not sure if that alone would easily allow us to have the kind of visualization this PR demonstrates. Does go-kit allow that? I'm not really familiar with it beyond exposing Prometheus metrics.

**Comment by mihaitodor** [comment]
@timoreimann Indeed, I could extend the StatsRecorder middleware to output the information that I need, but please note that the mutex that it currently uses [is locked only on the error path](https://github.com/containous/traefik/blob/2ddae2e856379f7ce2fb9b6509712e1ad523deeb/middlewares/stats.go#L83-L85). For my needs, I'll have to lock it for every incoming request.

> If not or when in doubt about the performance aspect, I'd follow Knuth's advice and not worry too much about the performance aspect for now. 
OK, cool! I don't mind doing that :)

> Re: go-kit/Prometheus: I'm not sure if that alone would easily allow us to have the kind of visualization this PR demonstrates.
I'm not at all familiar with Prometheus, although I've seen that it does expose some stats as well, but they seem aggregated. Also, I'd like this feature to be available for people who aren't using Prometheus.

Let me know if you agree with the above @timoreimann @emilevauge and I'll proceed to modify the StatsRecorder middleware and submit a PR.

**Comment by timoreimann** [comment]
I'm okay on the suggested StatsRecorder middleware changes.

Will let @emilevauge clarify if go-kit can help us out here somehow from his perspective.

**Comment by emilevauge** [comment]
@mihaitodor @timoreimann I'm OK to reuse the actual `StatsRecorder` middleware. But I'm a bit concerned with adding a mutex on the whole path (not only error). Couldn't we imagine a solution without mutex? Using channels and go routines for example ?

**Comment by timoreimann** [comment]
The alternative to the mutex would likely be a dedicated goroutine that periodically receives an updated value over a channel. Whether that will be more efficient than direct locking mostly depends on how much work the goroutine has to do: IIRC, pushing something into a channel involves locking anyway (as you need to make sure that enqueued data is consistent); and since channels / goroutines have a bigger overhead overall, the costs need to pay off on the separate goroutine. If it's just about increasing a counter, they likely won't. If there's more to do, it could be worth it.

Chances are I might miss a more simple approach involving goroutines. Unless we're super clear on what's going to be the cheaper solution, how about letting the owner of the feature implement both techniques, measure, and compare? It shouldn't be too hard to write the code for both ways. We may even add a little test benchmark function (which stdlib's `testing` package supports) to enable comparison.

WDYT?

**Comment by emilevauge** [comment]
I agree @timoreimann.
@mihaitodor are you OK with this ?

**Comment by mihaitodor** [comment]
@emilevauge I agree with writing some benchmarks and the 2 versions described above (goroutine and mutex for the StatsRecorder middleware). Also, as an additional alternative to having a mutex, I could try using the `sync/atomic` package [like so](https://gobyexample.com/atomic-counters), which, hopefully, should perform better than the mutex.

**Comment by timoreimann** [comment]
@mihaitodor atomic performing better than mutex would be my assumption too.

so we have a plan. :-)

**Comment by nmengin** [comment]
Hello there,

This proposal has been open for a while with minimal traction. We won't be adding it to our roadmap at this time and prefer to close it for now.

If circumstances change or interest grows, we can always reopen this later.
Thank you, @mihaitodor, for the suggestion!
